### PR TITLE
Restructuring pass of core spec information (IDP and Browser APIs)

### PR DESCRIPTION
**Pull request by bvandersloot-mozilla**
This should improve the clarity of these central sections of the spec and call out a few shortcomings

This should address the following issues from Issue #416: 7-9, 12, 13, 15. This leaves only 16 & 19 from issue #416 remaining.

### CODE REVIEW

**Review by npm1 — COMMENTED** [review]
Thanks for working on this! I think I don't love the placement of the dictionaries because there is no context on how they'd be used. I think they belong better closer to where they were. It might still be ok to have the order of the sections swapped if that is your preference, although that will mean the dictionaries are defined after they are used in the processing model.

**Comment by bvandersloot-mozilla** [comment]
> Thanks for working on this! I think I don't love the placement of the dictionaries because there is no context on how they'd be used. I think they belong better closer to where they were. It might still be ok to have the order of the sections swapped if that is your preference, although that will mean the dictionaries are defined after they are used in the processing model.

Maybe. I don't mind having the definition apart from its use. My concern is that the IDL definitions are normative. They describe the JSON structure required and expected. So I'd really like them to be in a normative section.

File: spec/index.bs [hunk 1]
```diff
@@ -265,1088 +265,1073 @@ identity federation.
...
+{{IdentityProviderAccount}} or failure.
+    1. Assert |accountsList|'s [=list/size=] is greater than 1.
+    1. Display an account chooser displaying the options from |accountsList|.
+    1. Let |account| be the {{IdentityProviderAccount}} of the account that the user
+        manually selects from the accounts chooser, or failure if no account is selected.
+    1. Return |account|.
+</div>
 
-TODO: add an ASCII image to explain how the states fit into the algorithms.
+Issue: Displaying an account chooser displaying all of the provided options should not be included normatively
```
[the remark this answers is not in the document]
Comment (bvandersloot-mozilla) [reply spec/index.bs:952]: Generally we shouldn't dictate that particular UI get displayed because a browser could obtain user choice can be rendered in multiple ways... e.g. advance configuration, an "always use this account" checkbox, or defaulting to choose the only account if only one is available. 

This spec has to toe a fine line between requiring some user consent/information without dictating UI/UX choices. and I am just pointing out that this is probably over that line IMO

File: spec/index.bs [hunk 2]
```diff
@@ -265,1088 +265,1073 @@ identity federation.
...
-<dl>
-    :   <b>{{IdentityProviderConfig/configURL}}</b>
-    ::  The URL of the configuration file for the identity provider.
-    :   <b>{{IdentityProviderConfig/clientId}}</b>
-    ::  The {{id_assertion_endpoint_request/client_id}} provided to the [=RP=] out of band by the [=IDP=]
-    :   <b>{{IdentityProviderConfig/nonce}}</b>
-    ::  A random number of the choice of the [=RP=]. It is generally used to associate a client
-        session with a {{IdentityProviderToken/token}} and to mitigate replay attacks. Therefore, this value should have
-        sufficient entropy such that it would be hard to guess.
+The file is parsed expecting a {{IdentityProviderWellKnown}} JSON object with the following semantics:
```
[the remark this answers is not in the document]
Comment (bvandersloot-mozilla) [reply spec/index.bs:1015]: unifying convo here: https://github.com/fedidcg/FedCM/pull/424#issuecomment-1424359330

File: spec/index.bs [hunk 3]
```diff
@@ -265,1088 +265,1073 @@ identity federation.
...
+it will not be used if the user agent does not find it in the well-known file. This allows the [=IDP=]
+to keep their actual config files on an arbitary path while allowing the user agent to prevent config file
+path manipulation to fingerprint (for instance, by including the RP in the path). See
+[[#manifest-fingerprinting]].
 
 <xmp class="idl">
-dictionary IdentityProviderClientMetadata {
-  USVString privacy_policy_url;
-  USVString terms_of_service_url;
+dictionary IdentityProviderWellKnown {
```
[the remark this answers is not in the document]
Comment (bvandersloot-mozilla) [reply spec/index.bs:654]: unifying convo here: https://github.com/fedidcg/FedCM/pull/424#issuecomment-1424359330

**Comment by npm1** [comment]
> Maybe. I don't mind having the definition apart from its use. My concern is that the IDL definitions are normative. They describe the JSON structure required and expected. So I'd really like them to be in a normative section.

I don't know about those IDL definitions being normative here. They are describing the expected format of the JSON file from IDP and are helpful as dictonaries because we can just quote them in the algorithm, but they are not really webexposed in the IDL sense so I think that those would be fine in the non-normative section.

Perhaps we can keep them where they are (although preference to move them to before their first use, and that does not seem to be their location right now) but add intro paragraphs to explain what we're defining? Right now there is no context to those IDLs

**Comment by bvandersloot-mozilla** [comment]
They end up being normative in a few ways.  For one, `IdentityProviderConfig` defines the structure of the object here: `navigator.credentials.get({"identity":{"providers":[HERE]}})`. Also, these objects are [converted](https://webidl.spec.whatwg.org/#dfn-convert-ecmascript-to-idl-value) from JSON objects in the "fetch the X file" algorithms, so changing those definitions changes the behavior of those algorithms. 

Although I definitely agree, there could be more context for their definition. I just put them in the list of defined algorithms after their first use, which seemed to be the convention followed in [that section](https://fedidcg.github.io/FedCM/#browser-api-rp-sign-in).

**Comment by npm1** [comment]
> Although I definitely agree, there could be more context for their definition. I just put them in the list of defined algorithms after their first use, which seemed to be the convention followed in [that section](https://fedidcg.github.io/FedCM/#browser-api-rp-sign-in).

If you mean the algorithms then yea I think that makes sense? For dictionaries, it should be easier and clearer to explain what they are and define them before they are used in the algorithms.

**Comment by bvandersloot-mozilla** [comment]
>If you mean the algorithms then yea I think that makes sense?

I mean to say that the algorithms and dictionaries are serving the same purpose here: to precisely specify what is required of an implementer in digestible pieces. The dictionary's IDL serves the purpose of defining the requirements of the structure (including the details of de-serializing it). So perhaps both would benefit from being explained more clearly.

> For dictionaries, it should be easier and clearer to explain what they are and define them before they are used in the algorithms.

Maybe? But in the Browser API section you only really need to know how they are used, not what they are and why. If the reader is confused by what the semantics of the values are, they can use [bikeshed links like this](https://pr-preview.s3.amazonaws.com/bvandersloot-mozilla/FedCM/pull/424.html#ref-for-dictdef-identityprovidericon%E2%91%A0) to navigate around for more context. Alternatively, this would be a great avenue for editorial improvement to the Introduction- especially giving context to some of those typograms at the end.

**Comment by npm1** [comment]
Ok, I think I'm ok with this. We should rebase though since there are merge conflicts. I also left some other comments but they can be resolved in a separate PR since this is mainly just reordering things.

**Comment by bvandersloot-mozilla** [comment]
Okay, I think I have resolved merge conflicts well. Sounds good!

File: spec/index.bs [hunk 4]
```diff
@@ -265,1085 +265,1070 @@ identity federation.
...
+    1. Let |account| be the {{IdentityProviderAccount}} of the account that the user
+        manually selects from the accounts chooser, or failure if no account is selected.
+    1. Return |account|.
+</div>
+
+Issue: Displaying an account chooser displaying all of the provided options should not be included normatively
 
 <!-- ============================================================ -->
-## The IdentityCredential Interface ## {#browser-api-identity-credential-interface}
+# Identity Provider Server Endpoints # {#idp-api}
```
Comment (samuelgoto) [inline spec/index.bs:951]:
```suggestion
# Identity Provider HTTP API # {#idp-api}
```

"Server Endpoints" isn't sitting really well with me ... can we call this "HTTP API" instead?

File: spec/index.bs [hunk 5]
```diff
@@ -265,1088 +265,1073 @@ identity federation.
...
-network requests. It is important for the browser
-to execute these in such a way that it does not allow the user to be tracked (by an attacker
-impersonating an [=IDP=]) on to the site using FedCM. The following table has information about the
-network requests performed:
+<div class=example>
+```js
+const credential = await navigator.credentials.get({
+  identity: {
+    providers: [{
+      configURL: "https://idp.example/manifest.json",
```
[the remark this answers is not in the document]
Comment (samuelgoto) [reply spec/index.bs:289]: There is a section called "Manifest" at the moment, which we should rename to "Config". I'd suggest to make this PR narrow and only about moving things around, and we can take a separate pass at cleaning up the Manifest -> Config renaming on a separate PR.

Fine to keep this change in this PR too, but my intuition is that we should make this as narrow as possible, so that we try to catch copy/pasting errors.

File: spec/index.bs [hunk 6]
```diff
@@ -265,1085 +265,1070 @@ identity federation.
...
-(a) **without** cookies,
-(b) **with** the <a http-header>Sec-Fetch-Dest</a> header set to `webidentity`, and
-(c) **without** revealing the [=RP=] in the <a http-header>Origin</a> or
-    [[RFC9110#field.referer|Referer]] headers.
+<dl dfn-type="argument" dfn-for="AccountState">
+    :   <dfn>registration state</dfn>
+    ::  Keeps track of whether the user agent is aware that the user has registered an account on the
+        [=RP=] with the [=IDP=] or not.
+         Can be <dfn>registered</dfn> or <dfn>unregistered</dfn> (by default).
+    :   <dfn>allows logout</dfn>
```
Comment (samuelgoto) [inline spec/index.bs:320]: Just FYI, we shouldn't address this in this PR, but this specific state here I think is only useful / used for the Front Channel Logout API, so we should probably move it in a separate PR to the front-channel-logout explainer / BS.

File: spec/index.bs (hunk 1 above)
[the remark this answers is not in the document]
Comment (samuelgoto) [reply spec/index.bs:948]: There is a lot going on in this PR and I'm a bit worried that it is going to be hard to determine / review what's just moving things around vs new text that we are introducing. For example, I missed this entirely, and if it weren't for npm double checking this would've probably slipped through. I don't know what the merits are of this specific point here being discussed, but a bit of a meta point is that I think it would be useful to try to make some of the moves so that we can easily tell (by looking at the diff) that we are keeping the normative algorithms / interfaces / etc intact without having to look at every single LOC.

I don't know if that's possible, but just wanted to throw it out there in this review that I worry about the scope of this PR.

**Review by samuelgoto — COMMENTED** [review]
I went through this PR and am a bit worried that it is hard for me to be confident that we are keeping the normative parts intact (I'm assuming that was the intent?). because we changed the orders of the sections (which sounds right to me), between the browser API and the IDP HTTP API, the diff is hard to parse for errors. I don't feel strongly about this, because we can just look more manually for copying/pasting errors, but if you can think of ways to make the move a bit easier to detect for errors, that'd be great.

**Comment by bvandersloot-mozilla** [comment]
So I split it into 6 commits and left out the Issue that seemed controversial. 

Only the first commit ([1fe4569](https://github.com/fedidcg/FedCM/pull/424/commits/1fe4569f7d8041870587e7b86e21e5ff9822c011)) is still hard to read. However if you check it out and do the diff with `--diff-algorithm=minimal`, you get a much cleaner diff (attached): [1fe4569.patch](https://github.com/fedidcg/FedCM/files/10723041/1fe4569.patch)

File: spec/index.bs (hunk 5 above)
[the remark this answers is not in the document]
Comment (samuelgoto) [reply spec/index.bs:288]: Resolving this since you did break this into separate commits.

File: spec/index.bs (hunk 6 above)
Comment (samuelgoto) [reply spec/index.bs:319]: Resolving this too.

File: spec/index.bs [hunk 7]
```diff
@@ -906,20 +906,12 @@ To <dfn>select an account</dfn> given an |accountsList|, run the following steps
...
 by exposing a series of HTTP endpoints:
 
+1. [[#idp-api-well-known]] that points to a Manifest
 1. A [[#idp-api-manifest]] endpoint in an agreed upon location that points to
 1. An [[#idp-api-accounts-endpoint]] endpoint
 1. A [[#idp-api-client-id-metadata-endpoint]] endpoint
 1. An [[#idp-api-id-assertion-endpoint]] endpoint
 
-
-The [=IDP=] must also host a file at the ".well-known/web-identity" path of its [=registrable domain=] that has JSON contents that are convertable to an {{IdentityProviderWellKnown}} object.
```
Comment (samuelgoto) [inline spec/index.bs]: We are losing this non-normative text here about the location of the ".well-known/web-identity" file. We are keeping the normative section, so, not a substantative change, but seemed worth noting that this is going to be lost in your move here.

**Review by samuelgoto — APPROVED** [review]
> So I split it into 6 commits and left out the Issue that seemed controversial.

This indeed helped review the patch. I did check manually the first commit and couldn't find anything else missing. 

I'm going to merge this since (a) this is a good restructuring that is going to help going forward, (b) the intent is to keep it purely non-normative, so in case we do find that this caused normative changes we can look at this history here to see where things may have lost, (c) there are other PRs that depend on this and (d) I think future PRs are going to be more incremental than this.
